### PR TITLE
259 email notifications

### PR DESCRIPTION
**Pull request by mimmon**
There is one fully prepared migration where we can put link to the order. Just wanted to make sure what should be in the link, is there any helper method or should I create it from regex in https://github.com/pyconsk/django-konfera/blob/master/payments/urls.py#L7  (something like `url = reverse('payment_options')` )?

There is also a function in payment utils that can be used as the next step - after several unsuccessful attempts no other email will be sent and order will be Cancelled - can be a separate Issue or I can do it in this issue.

### CODE REVIEW

File: konfera/settings.py [hunk 1]
```diff
@@ -53,8 +53,12 @@
...
+How many times we should repeat the email notification
+"""
+UNPAID_ORDER_NOTIFICATION_REPEAT = 3
 
 """
-Number of days after the user is notified about unpaid order
+How long shoudl we wait to notify about missing payment
 """
-UNPAID_ORDER_NOTIFICATION_DAYS = getattr(settings, 'UNPAID_ORDER_NOTIFICATION_DAYS', 3)
+UNPAID_ORDER_NOTIFICATION_REPEAT_DELAY = 5
```
Comment (ZoltanOnody) [inline konfera/settings.py]: It has to be over-writable in settings.

File: konfera/settings.py [hunk 2]
```diff
@@ -53,8 +53,12 @@
...
 EMAIL_NOTIFY_BCC = getattr(settings, 'EMAIL_NOTIFY_BCC', [])
 
+"""
+How many times we should repeat the email notification
+"""
+UNPAID_ORDER_NOTIFICATION_REPEAT = 3
 
 """
-Number of days after the user is notified about unpaid order
+How long shoudl we wait to notify about missing payment
```
Comment (ZoltanOnody) [inline konfera/settings.py]: typo in *should*

File: konfera/settings.py [hunk 3]
```diff
@@ -53,8 +53,12 @@
 """
 EMAIL_NOTIFY_BCC = getattr(settings, 'EMAIL_NOTIFY_BCC', [])
 
+"""
+How many times we should repeat the email notification
+"""
+UNPAID_ORDER_NOTIFICATION_REPEAT = 3
```
Comment (ZoltanOnody) [inline konfera/settings.py]: It has to be over-writable in settings.

File: payments/utils.py [hunk 4]
```diff
@@ -175,14 +176,20 @@ def check_payments_status(verbose=0):
             _process_payment(order, payment, verbose)
 
 
+def get_unpaid_orders(overdue=False):
+    deadline = timezone.now() - timedelta(days=UNPAID_ORDER_NOTIFICATION_REPEAT_DELAY)
+    orders = Order.objects.filter(Q(status=Order.AWAITING) | Q(status=Order.PARTLY_PAID))
+    orders = orders.filter(Q(date_created__lt=deadline) | Q(unpaid_notification_sent_at__lt=deadline))
+    if overdue:
```
Comment (ZoltanOnody) [inline payments/utils.py]: What is `overdue` and why is it here if we don't use it anywhere?

File: payments/utils.py [hunk 5]
```diff
@@ -175,14 +176,20 @@ def check_payments_status(verbose=0):
             _process_payment(order, payment, verbose)
 
 
+def get_unpaid_orders(overdue=False):
+    deadline = timezone.now() - timedelta(days=UNPAID_ORDER_NOTIFICATION_REPEAT_DELAY)
+    orders = Order.objects.filter(Q(status=Order.AWAITING) | Q(status=Order.PARTLY_PAID))
+    orders = orders.filter(Q(date_created__lt=deadline) | Q(unpaid_notification_sent_at__lt=deadline))
```
Comment (ZoltanOnody) [inline payments/utils.py]: Hmm... I am not sure this query does what we want.

I think we need something like:
```
(Q(date_created__lt=deadline) & Q(unpaid_notification_sent_amount=0)) # we haven't sent any notification yet 
| Q(unpaid_notification_sent_at__lt=deadline)  # or we sent the notification X days ago.
```

Comment (mimmon) [inline payments/utils.py]: You're right, the original query will actually ignore the notification date.

File: payments/utils.py (hunk 4 above)
Comment (mimmon) [inline payments/utils.py]: Overdue means it has not been paid within given period of time (i.e. after three notifications). So it is  just a preparation for cancelling unpaid orders.

File: payments/utils.py [hunk 6]
```diff
@@ -175,14 +176,21 @@ def check_payments_status(verbose=0):
             _process_payment(order, payment, verbose)
 
 
+def get_unpaid_orders(overdue=False):
+    deadline = timezone.now() - timedelta(days=UNPAID_ORDER_NOTIFICATION_REPEAT_DELAY)
+    orders = Order.objects.filter(Q(status=Order.AWAITING) | Q(status=Order.PARTLY_PAID))
+    orders = orders.filter(Q(unpaid_notification_sent_amount=0) & Q(date_created__lt=deadline) |
```
Comment (ZoltanOnody) [inline payments/utils.py]: I am not sure whether this means (A and B) or C  ...or... A and (B or C). Could you put there the parenthesis explicitly?

File: payments/utils.py (hunk 4 above)
Comment (mimmon) [inline payments/utils.py]: I will do it, no problem. 
I also want to change the email template as requested in specs of the issue.

**Comment by mimmon** [comment]
There is still some work to be done - the email template (in migration) should be changed to accept link to order and then add link in the code.

**Comment by ZoltanOnody** [comment]
Okay, that's why the author of the PR should merge it. I just approved the changes because they seems to be OK. If there are any work in PR left, you can still commit them. :)

File: konfera/settings.py [hunk 7]
```diff
@@ -70,3 +70,14 @@
 How long should we wait to notify about missing payment
 """
 UNPAID_ORDER_NOTIFICATION_REPEAT_DELAY = getattr(settings, 'UNPAID_ORDER_NOTIFICATION_REPEAT_DELAY', 5)
+
+"""
+Absolute url base with protocol
+"""
+ABSOLUTE_URL = getattr(settings, 'ABSOLUTE_URL', 'https://www.pycon.sk/')
```
Comment (ricco386) [inline konfera/settings.py]: How about we rename it to SITE_URL, you can leave in comment that it is absolute url. I can imagine we will need SITE_NAME for the emails, possibly others as well.

File: konfera/settings.py [hunk 8]
```diff
@@ -70,3 +70,14 @@
 How long should we wait to notify about missing payment
 """
 UNPAID_ORDER_NOTIFICATION_REPEAT_DELAY = getattr(settings, 'UNPAID_ORDER_NOTIFICATION_REPEAT_DELAY', 5)
+
+"""
+Absolute url base with protocol
```
Comment (ricco386) [inline konfera/settings.py]: please add a note it always have to have end without (once we use reverse) trailing slash (/) so it is explicit.

File: payments/utils.py [hunk 9]
```diff
@@ -176,6 +176,12 @@ def check_payments_status(verbose=0):
             _process_payment(order, payment, verbose)
 
 
+def get_full_order_url(order):
+    ABSOLUTE_URL = settings.get('ABSOLUTE_URL')
+    full_url = '{}order/{}'.format(ABSOLUTE_URL, order.uuid)
```
Comment (ricco386) [inline payments/utils.py]: please use [reverse](https://docs.djangoproject.com/en/1.10/ref/urlresolvers/#reverse) this wont work if we change url for some reason...

File: konfera/settings.py (hunk 8 above)
Comment (mimmon) [inline konfera/settings.py]: I will add it. I thought about testing it in code just for sure, but it is probably cheaper to get error soon and have it fixed than check every single time if it is in required format. Am I right?

**Pull request merged**